### PR TITLE
feat(risingwave): add streaming DDLs

### PR DESCRIPTION
**Pull request by KeXiangWang**
## Description of changes

Risingwave plays a role as a computing engine and storage system in Streaming ecosystems.
Usually, a streaming workload will include two/three systems like this:
`upstream data source --> Risingwave`
Or,
`upstream data source --> Risingwave --> downstream data sink`

This PR adds streaming support, mainly new relations related to streaming, for the Risingwave backend.

To be specific, Four types of new relations are introduced:
- `Source`, data sources that encompass a connector connected to an upstream data system like Kafka. A source works like an extraction or a placeholder for an upstream system, and although it has columns, it does not store any data itself.
- `Materialized View`, which is Risingwave's core concept for streaming. One can create an MV with a query on existing tables or sources. The data in MV is automatically updated in a real-time way. Users can access the data with a `select` statement just like accessing a table.
- `Table` with connector, works like a combination of a source and a normal table. The difference between `Table` with connector and `Source` is that, once the table is created, it will automatically start to consume data from upstream systems and update it into Risingwave.
- `Sink`, unlink a `Materialized View` which stores the result of a query in RW, users can choose to sink the result to a downstream system, e.g. Redis, and then read the data in the downstream system. Unlike an MV, User cannot access the data directly from a sink.

Some minor changes:
1. Pump the image version to a new nightly build, as the new image solves some issues found in the [previous PR](https://github.com/ibis-project/ibis/pull/7954).
2. Pump the `sqlalchemy-risingwave` version from `1.0.0` to `1.0.1`, some implementations are updated.
3. Mark one test, `test_semi_join`, as skipped for risingwave backend. The test sometimes stuck and it seems to be Risingwave's fault. I'll continue to investigate.

One issue introduced in this PR:
Risingwave backend's implementation is sqlalchemy-based. But sqlalchemy has no corresponding concept for sources and sinks. So, in order to work around this issue, I temporarily categorize a source as a view. So sqlalchemy can access a source's metadata (names and types of its columns) like a view. However, this causes a side effect, when users call `list_view`, they may see some unexpected views that are actually sources. This can be fixed by rewriting the implementation of the `list_view` func in the Risingwave Backend. I'll fix it later.

### CODE REVIEW

**Comment by KeXiangWang** [comment]
It seems I have no access to invite reviewers. @gforsyth @cpcloud @chloeh13q @deepyaman Could you please help take a review or invite the appropriate ones?

**Comment by jcrist** [comment]
Hi @KeXiangWang - thanks for opening this! 

We're unfortunately in the midst of a large internals refactor moving all backends to use `sqlglot` instead of `sqlalchemy`. We hope to merge https://github.com/ibis-project/ibis/tree/the-epic-split into `main` soon (maybe by end of week?) after which we'll be back to accepting PRs to `main`. Until that's done though we don't really want to merge additional large PRs as this would complicate getting the refactor in. We'll follow up after `the-epic-split` is merged to discuss how best to implement the work here using `sqlglot`, but until then things are blocked on us finishing up the internals refactor.

**Comment by zhenzhongxu** [comment]
@chloeh13q @cpcloud: bump this up, is this something we can help wth getting the sqlglot refactoring?

**Comment by KeXiangWang** [comment]
> It looks like the first order of business would be to rebase this PR on main.

OK, I'll try my best.

**Comment by KeXiangWang** [comment]
Hi @cpcloud, I've rebased the PR to newest main and fix all the issue introduced by the `sqlglot` refactoring. Could you please help take a look.

Besides, I find some tests are unable to pass even with the main branch, so I leave them unchanged for now. Do you have any ideas? Is there any configuration I missed?

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:568]:
```suggestion
            The table or materialized view name to sink from. Only one of `sink_from` or `obj` can be
```

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:674]: You might as well combine these into a single method if the first is only doing a check for `None`

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:672]: Do these need to be sorted?  If not, I think we should render them in the order they were provided.

**Review by gforsyth — COMMENTED** [review]
Hey @KeXiangWang -- thanks for putting this together and sorry for the delay in reviewing!

I've taken first pass over the new DDL stuff.  I am not a streaming expert, so maybe @chloeh13q and @deepyaman can also chime in on some of the API design stuff.

Since some of these methods may become "standard" on other (future or current) streaming backends we want to be very deliberate in how we design them.  We especially want to make sure that we don't overindex on a single backend when coming up with names (this is very difficult).

File: ibis/backends/risingwave/__init__.py
Comment (mfatihaktas) [inline ibis/backends/risingwave/__init__.py:399]: Similar to other comments, can we not use `drop_materialized_view()` for this?

File: ibis/backends/risingwave/__init__.py
Comment (chloeh13q) [reply ibis/backends/risingwave/__init__.py:352]: I think this information may be helpful in helping us decide on a suitable API: https://materialize.com/guides/materialized-views/#how-do-materialized-views-work-in-specific-databases

Personally I think it depends on whether people tend to think of materialized views as a special type of views, or completely separate from normal views. It sounds like several backends just treat materialized views as a special type of views, as opposed to "its own thing".

- Postgres: https://www.postgresql.org/docs/current/rules-materializedviews.html
- Materialize: https://materialize.com/guides/materialized-views/#materialized-view-example

Comment (chloeh13q) [reply ibis/backends/risingwave/__init__.py:352]: Also wondering: 

I saw that there is a ticket for implementing materialized views for a broader set of backends: https://github.com/ibis-project/ibis/issues/5964 - is this an API we'd want to include in the base class?

**Comment by chloeh13q** [comment]
I have some open questions on a high level:

I understand the motivation behind introducing separate APIs for sources, materialized views, tables, and sinks for the RW backend. Sources, materialized views, tables, and sinks are distinct objects in RW. However, I wonder how generalizable this is across different streaming backends.

For the user who is already familiar with RW's API, this implementation feels natural. But for the user who is coming from another backend, it may cause some confusion.

For example, Flink also has the same concepts, but it doesn't use a different API for each. Source, tables, and sinks are all created with `CREATE TABLE`. A lot of these underlying differences are abstracted away from the user.

Of course, it's also okay that we just do this for the RW backend for now and refactor at a later point if there is a need to consolidate some of the APIs.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:352]: It's easy to make `materialized view` a special type if `view` in this PR. But I would say they should be different objects in databases. The underlying mechanism and performance are largely different. For streaming databases like Risingwave and Materialize, `Materialized view` is the core concept and hardest part to design, while `view` is much more trivial. So we would like to keep `materialized view` and `view` separate.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: Got it. Updated.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: updated it to tbl_properties

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: Now, they are `tbl_properties ` and `encode_properties `.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: Tried to add more description..

Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: Updated, now they are at the end of the kward-list.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:451]: Depends on how we consider `materializedv view` vs `view.` Personally, I'm inclined to keep `drop view` and `drop materialized view` as two independent functions if we have `create view` and `create materialized view` separate.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: updated.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: updated

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: `sorted` removed

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:411]: 👀 Do you mean we should use `drop view`? Or you think we should use `drop_materialized_view()` here.

**Comment by KeXiangWang** [comment]
> I wonder how generalizable this is across different streaming backends.
Good question. These concepts are common in streaming backends, although different systems may have different names for them. For example, as Flink doesn't have a normal persistent table in a traditional database, its tables are used to express a streaming job. While in RW, Materialize, and Timeplus, we use `materialized views`. For sources, Flink combines a streaming job and the dependent sources together as a Table, while RW and materialize use `sources` objects and Timeplus use `stream` objects.  
> Flink also has the same concepts, but it doesn't use a different API for each.
Flink positions itself as a 'compute engine,' which means it's designed not to store data but to process it. In Flink, a `table` is essentially a streaming job that interacts with sources or sinks. The outcomes of these jobs are not stored within Flink itself. In contrast, streaming databases such as RW, Materialize, and Timeplus, function as databases that store the results locally. While Flink tables are declared as `Tables`, you cannot directly query data from them. Instead, you must route the data to an external system to access the results. With Flink, you trade off a simpler API for the flexibility and simplicity of its architecture. On the other hand, streaming databases offer the ability to customize your streaming jobs by combining various objects to suit your needs, providing direct accessing or sinking to external systems at the same time.

File: ibis/backends/risingwave/__init__.py
Comment (zhenzhongxu) [reply ibis/backends/risingwave/__init__.py:364]: Do we also want to distinguish between incremental materialized view?

File: ibis/backends/risingwave/__init__.py
Comment (mfatihaktas) [reply ibis/backends/risingwave/__init__.py:411]: I meant using the appropriate function, ``drop_view()`` or `drop_materialized_view()`, instead of the `sqlglot` expression:
```
sge.Drop(kind="MATERIALIZED VIEW", this=this, exists=True).sql(
                        self.dialect
                    )
```
IMO, `sqlglot` expressions look complex and replacing them with calls to the appropriate functions would improve readability.

Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: got it. updated

File: ibis/backends/risingwave/__init__.py
Comment (cpcloud) [reply ibis/backends/risingwave/__init__.py:362]: I don't think we should distinguish here without a clear explanation of what the difference between incremental and non-incremental streaming materialized views are.

It's also adding more scope than is needed for now.

Comment (cpcloud) [reply ibis/backends/risingwave/__init__.py:362]: I think we should keep this API separate for now, it's likely that materialized views will require some additional kwargs that don't apply to views. The word "view" here is really being stretched to its limit by this functionality and thinking about materialized views as another kind of view isn't really correct.

File: ibis/backends/risingwave/__init__.py
Comment (cpcloud) [inline ibis/backends/risingwave/__init__.py:165]: Are all of these nouns the same ones used in the RisingWave documentation? If not, we should make them match and avoid introducing new nouns just for the Ibis API.

File: ibis/backends/risingwave/__init__.py
Comment (cpcloud) [inline ibis/backends/risingwave/__init__.py:48]: Is all this validation necessary? What if a new data format or encoding is added? Can't we let this fall through to the driver to avoid kneecapping users who upgrade their risingwave installation?

File: ibis/backends/risingwave/__init__.py
Comment (cpcloud) [inline ibis/backends/risingwave/__init__.py:350]: Why do you need this additional method? Try to avoid adding public APIs like this.

**Comment by chloeh13q** [comment]
@KeXiangWang Yep that makes sense. Since Ibis works across engines & databases, I'd imagine that these are questions that Ibis will need to address. But I think it also depends on the background of the user, e.g., whether it's someone who's experimenting with streaming or someone who already. has some expertise in streaming and wants to try out the Ibis API. In any case, these may be questions that we can leave open and come back to, when there is user feedback backing up one way or another.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:165]: It would be more natural, if we replace `tbl_properties` with `connector_properties`, but it will differentiate from Flink's corresponding nouns.

Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:165]: Do you think it's appropreciate to rename it to `connector_properties`?

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:48]: Make sense. removed.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:350]: Sorry, it should have been deleted. Removed just now.

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [reply ibis/backends/risingwave/__init__.py:134]: If that matches the Risingwave terminology, then I think that is a good naming choice.

**Comment by gforsyth** [comment]
@KeXiangWang -- we've just merged in #8655 which moves us away from using the word `schema` in any hierarchical sense.  We're standardizing on "database" as a collection of tables and "catalog" as a collection of database.
If this PR allows maintainers to push updates, I can take on porting over your work here to the new naming and push that up.

**Comment by KeXiangWang** [comment]
> If this PR allows maintainers to push updates, I can take on porting over your work here to the new naming and push that up.

Thx @gforsyth . I really cannot fiind a button in this page to enable it, so I create a new PR [here](https://github.com/ibis-project/ibis/pull/8781). We can push the following updates to the new PR. Thanks!

**Comment by gforsyth** [comment]
I've pushed up the fixes to #8781 -- @KeXiangWang, you can either cherry-pick that commit back to this PR, or we can close this PR out in favor of your newer one, whichever you prefer!

**Comment by KeXiangWang** [comment]
Hi @gforsyth . Thanks for the effort. I've update the patch to this PR and I'll close https://github.com/ibis-project/ibis/pull/8781 when this PR is merged. Thanks!

**Comment by KeXiangWang** [comment]
@cpcloud @gforsyth Any further comments? Can we merge this PR now?

File: ibis/backends/risingwave/__init__.py
Comment (cpcloud) [inline ibis/backends/risingwave/__init__.py:320]: Can you add `@experimental` here? This will indicate that the method may change between non major releases.

Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:322]: added.

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:179]:
```suggestion
            raise com.UnsupportedArgumentError(
```

I think this is more appropriate here, but not a blocker

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:520]: this should probably be pandas DataFrame and the type hint above should also include `pd.DataFrame` -- however, is this supported / tested?

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:533]:
```suggestion
            raise com.UnsupportedArgumentError(
```

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:536]: I think we should raise here if both `sink_from` and `obj` are provided -- currently this ignores `obj` silently if `sink_from` is provided.

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:532]: I think you want an `and` here -- as it's currently written I think this will error if both kwargs are `None`

```python
None is None != None is None
True != None is None
False is None
False
```

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [inline ibis/backends/risingwave/__init__.py:330]:
```suggestion
        """Create a materialized view. Materialized views can be accessed like a normal table.
```

small language nit

**Review by gforsyth — CHANGES_REQUESTED** [review]
Thanks for pushing on this @KeXiangWang !

I flagged a few style things and one bit of incorrect logic, I think.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: Yes, you are right. `pandas table` should not be added. It should be a typo. Updated

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: Updated. But kind of confused: what's the differences between `RelationError `, `UnsupportedArgumentError` and `ValueError`?

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py:540]: updated.

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: udpated

File: ibis/backends/risingwave/__init__.py
Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: updated

File: ibis/backends/risingwave/__init__.py
Comment (gforsyth) [reply ibis/backends/risingwave/__init__.py]: It's confusing!

`com.RelationError` should maybe go away, but currently it's only for internal errors related to expression construction -- no user should ever see a `RelationError`.

`ValueError` is (in my opinion) generally too vague to be useful, but it's tricky to balance adding a bunch of custom exceptions.

`UnsupportedArgumentError` seemed like a good fit to me, but there isn't a strict guide on when to use which exceptions in Ibis.  I think it's something we can improve for new contributors.

Comment (KeXiangWang) [reply ibis/backends/risingwave/__init__.py]: OKK, thanks for the explanation!
As there's no comments left, can we merge it now?

**Review by gforsyth — APPROVED** [review]
Thanks for all of your work on this @KeXiangWang !

**Pull request merged**
